### DISCUSSION TITLE
Where are Inlay Hints?

### DISCUSSION THREAD

**Discussion by ckoehler** [mrcjkb/rustaceanvim#46] [Q&A] [answered]
Don't see them; what am I missing?

**Comment by mrcjkb** [comment, marked as answer]
Hey :wave:

I removed the inlay hints code, because

- They will be a built-in feature of neovim 0.10.
- There exist plugins that provide the feature for neovim 0.9.

See [the changelog](https://github.com/mrcjkb/rustaceanvim/blob/master/CHANGELOG.md#breaking-changes-compared-to-rust-toolsnvim) for a list of breaking changes compared to rust-tools.nvim.

**Reply by liam-ngy** [reply]
How would enable it back again :) Right now I have to enable it via the lua command.

I tried this config

```
  {
    "mrcjkb/rustaceanvim",
    event = "BufReadPost",
    version = "^3", -- Recommended
    ft = { "rust" },
    config = function()
      vim.g.rustaceanvim = {
        -- Plugin configuration
        -- tools = {},
        -- LSP configuration
        server = {
          on_attach = function(client, bufnr)
            -- you can also put keymaps in here
            vim.lsp.inlay_hint(bufnr, true)
          end,
          --     settings = {
          --       -- rust-analyzer language server configuration
          --       ["rust-analyzer"] = {},
          --     },
          --   },
          --   -- DAP configuration
          --   dap = {},
        },
      }
    end,
  },

```

But so far it doesn't seem to work. Any ideas?

**Reply by liam-ngy** [reply]
Nevermind it works! But for some reason I don't like to see them anymore. They are kinda annoying

**Reply by mrcjkb** [reply]
Glad to hear you were able to solve it!
I personally like to keep them disabled by default and have a keymap (e.g. `<space>h`) to toggle thhem.

**Reply by benzwt** [reply]
I try your @liam-ngy config for both version 3 and 4 on NVIM v0.10.0-dev. But it doesn't work.
I got an error: 

>LSP[rust-analyzer]: Error ON_ATTACH_ERROR: "/xx/xxx/xx/plugin.lua": attempt to call field inlay_hint (a table value)

thanks!
would you share your workable solution ?

@mrcjkb I do like to a keymap to toggle the hints.
what would the config looks like ?

**Reply by mrcjkb** [reply]
It has changed to `vim.lsp.inlay_hint.enable(/ * boolean */)`

[This](https://github.com/mrcjkb/kickstart-nix.nvim/blob/6b28fa398a69b99318bde099fb9566eead5fa02e/nvim/plugin/autocommands.lua#L94) is the keymap I use.

**Reply by benzwt** [reply]
thanks! it is very helpful !

**Comment by fjchen7** [comment]
hi @mrcjkb Do you have the way to enable inlay hint? I try to integrate https://github.com/lvimuser/lsp-inlayhints.nvim and https://github.com/simrat39/inlay-hints.nvim in rust lsp `on_attach` method, but both fail to work. my neovim is 0.9

**Reply by mrcjkb** [reply]
Hey :wave: 

Do you have a snippet of (or ideally a URL to) your attempts so far?

**Reply by fjchen7** [reply]
Thanks for your help. I use lazy.nvim and here is my configuration:

```lua
{
  "mrcjkb/rustaceanvim",
  version = "^3",
  dependencies = {
    "nvim-lua/plenary.nvim",
    "mfussenegger/nvim-dap",
    {
      "lvimuser/lsp-inlayhints.nvim",
      opts = {}
    },
  },
  ft = { "rust" },
  config = function()
    vim.g.rustaceanvim = {
      inlay_hints = {
        highlight = "NonText",
      },
      tools = {
        hover_actions = {
          auto_focus = true,
        },
      },
      server = {
        on_attach = function(client, bufnr)
          require("lsp-inlayhints").on_attach(client, bufnr)
        end
      }
    }
  end
}

```

**Reply by mrcjkb** [reply]
:thinking: I just tried it out. I can confirm that the `on_attach` function is being called (can you, too? e.g. by adding `print('DEBUG')` to it?).

But in my test, I didn't see any inlay hints - which means it might be an issue with the lsp-inlayhints plugin?

**Reply by mrcjkb** [reply]
After adding `require("lsp-inlayhints").setup()` (I think lazy.nvim does that automatically?) and calling `require("lsp-inlayhints").show()`, or `require("lsp-inlayhints").toggle()` I can see inlay hints with the `lvimuser/lsp-inlayhints.nvim` plugin (after quite some delay).

If you want them to show up automatically, you should be able to call `require("lsp-inlayhints").show()` in the `on_attach` function.

As a side note:  The `vim.g.rustaceanvim.server.inlay_hints` config doesn't do anything, as the feature has been removed from this plugin. You can use the [lsp-inlayhints.nvim configuration](https://github.com/lvimuser/lsp-inlayhints.nvim#configuration) instead.

**Reply by jose-verissimo** [reply]
Takes quite a long time to load, does this happen to you as well? It's around 5 seconds every page open...

**Reply by mrcjkb** [reply]
Yes, it was rather slow when I tested the plugin.

**Reply by SacrilegeWasTaken** [reply]
Hey! How to return old beautiful inlay type hints? I tried configs from this discussion, and It works, but hints were ugly black squares instead of eye pleasant old hints that not hurting eyes. Maybe I did something wrong? Help please:(

**Reply by mrcjkb** [reply]
If you're using neovim nightly, with built-in inlay hints, it's possible your colorscheme doesn't support them yet. You can either request they define a highlight for the `LspInlayHint` highlight group, or change it yourself.

If you're using one of the plugins, please ask there 😅

**Reply by nicos68** [reply]
Hi, is this supposed to work with nvim 0.9 ?

I applied this configuration (see my config [here](https://codeberg.org/nise/dotfiles/src/branch/inlay-hints/.config/nvim/after/plugin/rustaceanvim.lua) for reference) but the hints are not displayed.

**Reply by mrcjkb** [reply]
@nicos68 rustaceanvim doesn't have a `inlay_hints` config. 

Please see the [lsp-inlayhint README section on highlights](https://github.com/lvimuser/lsp-inlayhints.nvim?tab=readme-ov-file#highlight) and [`:h nvim_set_hl`](https://neovim.io/doc/user/api.html#nvim_set_hl()) or [`:h highlight`](https://neovim.io/doc/user/syntax.html#%3Ahighlight).

**Reply by nicos68** [reply]
I basically just forgot to call the `setup` function of the `inlayhints` plugin 😑

**Comment by TomzBench** [comment]
i just updated to nvim 0.10 today and maybe `vim.lsp.inlay_hint` is moved somewhere else? it seems to error when i call it with attempt to call field 'inlay hint' (a table value)

**Reply by TomzBench** [reply]
yep it is now `vim.lsp.inlay_hint.enable(/* ... */)`

**Comment by LinoWhy** [comment]
Is it still possible to make the inlay_hints still set at the end of line? 
"Inlined" inlay_hints is too aggressive while coding.

**Reply by mrcjkb** [reply]
As it is a built-in neovim feature, you're probably better off asking there 😄

**Reply by PPakalns** [reply]
Created feature request.   https://github.com/neovim/neovim/issues/28261

I hope that this can be implemented, currently inlay hints are unusable.

**Reply by spenserlee** [reply]
I just found this plugin which restored the previous end of line hint I was used to with rust-tools:

https://github.com/chrisgrieser/nvim-lsp-endhints

The suggested configuration for lazy worked for me.
